### ISSUE TITLE
Refactor graph to support limited node metadata

### ISSUE THREAD

**Issue by decentralion** [sourcecred/sourcecred#1136]
updated plan
---

I intend to refactor the Graph so that rather than storing nodes only by identity (`NodeAddressT`), we have a `Node` type with a few fields, such as:

```
type Node = {|
  +address: NodeAddressT,
  +timestampMs: number | null,
  +description: string,
|}
```

This is motivated by the fact that a few extra fields (e.g. timestamp) are needed for generally computing cred, and packaging them together will greatly simplify data plumbing for SourceCred. See [this comment for justification and discussion of tradeoffs](https://github.com/sourcecred/sourcecred/issues/1136#issuecomment-498078988).

This involves a major refactor to the Graph class. In particular:
- We need to modify the Graph so nodes are stored as objects.
- We need to ensure that adding a node produces an error if the contents conflict (but is idempotent if the node is identical)
- We need to ensure that it is legal to add dangling edges (so that e.g. the GitHub plugin can add an edge to a Git node without having sufficient information to add it to the graph)
- We need to ensure that dangling edges are handled appropriately (e.g. they should be wholly ignored for PageRank purposes)
- We need to ensure that Graph.merge throws an error if node contents conflict
- We need to ensure that JSON serialization of node contents and of dangling edges work as expected

original issue
---
I'm interested in modifying the Graph class so that every node is stored along with a markdown description.

This would be convenient from the perspective of the current implementation. We could, for example, more-or-less remove the Dynamic Explorer adapter (its other function, providing the graph, is getting obsoleted by generating PagerankGraphs on the backend (#967)). It would also save us from writing new Graph wrappers that mostly serve to add descriptions to nodes, e.g. the OdysseyInstance in #1134.

Naturally, "it would be convenient right now" is not sufficient justification to add a field to the core data structure (more than just "add a field": it  increase the number of fields from 0 to 1). However, I do think that having a description built-in to the graph would be of general value. Any time that we want to be able to introspect on the graph, or debug or interpret scores, having info on what the nodes represent will be valuable. Also, using Markdown gives a lot of flexibility, since it supports hyperlinks, formatting, etc. 

Basically, having seen that the first two frontends to SourceCred (explorer and odyssey-ui) both need descriptions, I suspect that they all will, and if so, putting descriptions directly in the graph will make things a lot simpler and more consistent from an implementation perspective.

The downsides to this approach include:
- potential for enormous graph bloating
- waste in the cases where we don't need introspection (pure score calculation)
- slippery slope towards adding more and more fields to the graph

One might push back against this proposal by saying: didn't we spend a great deal of effort slimming down the graph to get rid of metadata? Why walk back those changes? I think the key difference here is that the previous major graph refactor removed arbitrary, plugin-provided metadata, which was un-manageably complicated (and implied re-inventing DBs in general). Adding a single field that is of cross-cutting relevance does not bring back those issues.

@wchargin, I'm eager for your thoughts.

**Comment by mzargham** [comment]
What if we looked for ways to decouple that data from the graph while retaining the mapping? A simple example would be to put the markdown description in IPFS and then store its hash in the graph so you can look it up.

**Comment by decentralion** [comment]
> What if we looked for ways to decouple that data from the graph while retaining the mapping? A simple example would be to put the markdown description in IPFS and then store its hash in the graph so you can look it up.

Yeah, having a 'GraphDescriptionStore' (or similar) associated with the Graph would also be a reasonable approach. It would accomplish my desire for there to be a 1st-class way to bundle descriptions with the graph (rather than doing it ad-hoc for every plugin), and would scale better to adding more kinds of data. (e.g. then we could also have a 'TypeStore' for storing type metadata, rather than also bundling it in the Graph).

**Comment by decentralion** [comment]
I've been continuing to ruminate on this for the past ~month, and I've decided to push ahead with this approach. A tipping point has been developing timeline cred; timestamps are now necessary to interpret and score the graph, so it's important that we have a consistent way to package timestamps with the graph. I've tried approaches like described in [the comment above](https://github.com/sourcecred/sourcecred/issues/1136#issuecomment-486375103); see https://github.com/sourcecred/sourcecred/pull/1162 for an idea of what this looks like. What I've found in practice is that keeping a constellation of related files that need to be kept in sync with the graph creates a lot of unnecessary friction, and I'd much prefer the simplicity of centralizing this information in the graph.

This looks suspiciously like our old type-paramerized graph, which we migrated away from with considerable effort (see #328 for some of that work). However, we are not allowing plugins to provide arbitrary payloads; it will be restricted to a few fields that are of extremely general value within SourceCred. (Right now, I imagine timestamp, description, and eventually NodeTye as the fields.) 

Once this change is fully implemented, I believe it will be possible to remove the adapters. Currently, adapters provide descriptions, timestamps, type information, and the graph. By simply building an aggregated graph in `sourcecred load`, no adapters will be needed. This will greatly simplify adding new plugins, and data pipelining in general.

In making this change, I am consciously trading certain kinds of flexibility for velocity. If we later have applications where the descriptions fields or timestamp field are irrelevant, and we need streamlined graphs, then this change may be inconvenient. However, I'm pretty confidence in the general usefulness of these fields, and am willing to accept the tradeoff.
